pkg/models/alerting/rules/ruler.go

```diff
@@ -102,38 +102,108 @@ func (r *ruleResource) deleteAlertingRules(rules ...*RuleWithGroup) (bool, error
 // If there are rules to be updated, return true to indicate the resource should be updated.
 func (r *ruleResource) updateAlertingRules(rules ...*RuleWithGroup) (bool, error) {
 	var (
-		commit  bool
-		spec    = r.Spec.DeepCopy()
-		ruleMap = make(map[string]*RuleWithGroup)
+		commit     bool
+		spec       = r.Spec.DeepCopy()
+		ruleMap    = make(map[string]*RuleWithGroup)
+		ruleGroups = make(map[string]map[string]struct{}) // mapping of name to group
 	)
 
 	if spec == nil {
 		return false, nil
 	}
+	spec.Groups = nil
 
 	for i, rule := range rules {
 		if rule != nil {
 			ruleMap[rule.Alert] = rules[i]
+			ruleGroups[rule.Alert] = make(map[string]struct{})
 		}
 	}
 
+	// Firstly delete the old rules
+	for _, g := range r.Spec.Groups {
+		var rules []promresourcesv1.Rule
+		for _, r := range g.Rules {
+			if r.Alert != "" {
+				if _, ok := ruleMap[r.Alert]; ok {
+					ruleGroups[r.Alert][g.Name] = struct{}{}
+					commit = true
+					continue
+				}
+			}
+			rules = append(rules, r)
+		}
+		if len(rules) > 0 {
+			spec.Groups = append(spec.Groups, promresourcesv1.RuleGroup{
+				Name:                    g.Name,
+				Interval:                g.Interval,
+				PartialResponseStrategy: g.PartialResponseStrategy,
+				Rules:                   rules,
+			})
+		}
+	}
+
+	// Then add the new rules
+	var groupMax = -1
 	for i, g := range spec.Groups {
-		for j, r := range g.Rules {
-			if r.Alert == "" {
+		if len(ruleMap) == 0 {
+			break
+		}
+
+		if strings.HasPrefix(g.Name, customRuleGroupDefaultPrefix) {
+			suf, err := strconv.Atoi(strings.TrimPrefix(g.Name, customRuleGroupDefaultPrefix))
+			if err != nil {
 				continue
 			}
-			if b, ok := ruleMap[r.Alert]; ok {
-				if b == nil {
-					spec.Groups[i].Rules = append(g.Rules[:j], g.Rules[j+1:]...)
-				} else {
-					spec.Groups[i].Rules[j] = b.Rule
-					ruleMap[r.Alert] = nil // clear to mark it updated
+			if suf > groupMax {
+				groupMax = suf
+			}
+		}
+
+		var num = customRuleGroupSize - len(g.Rules)
+		if num > 0 {
+			for name, rule := range ruleMap {
+				if num <= 0 {
+					break
+				}
+				if gNames, ok := ruleGroups[name]; ok {
+					// Add a rule to a different group than the group where it resided, to clear its alerts, etc.
+					// Because Prometheus may migrate information such as alerts from the old rule into the new rule
+					// when updating a rule within a group.
+					if _, ok := gNames[g.Name]; !ok {
+						spec.Groups[i].Rules = append(spec.Groups[i].Rules, rule.Rule)
+						num--
+						delete(ruleMap, name)
+						commit = true
+					}
 				}
-				commit = true
 			}
 		}
 	}
 
+	for groupMax++; len(ruleMap) > 0; groupMax++ {
+		g := promresourcesv1.RuleGroup{Name: fmt.Sprintf("%s%d", customRuleGroupDefaultPrefix, groupMax)}
+		var num = customRuleGroupSize
+		if num > 0 {
+			for name, rule := range ruleMap {
+				if num <= 0 {
```

Comment:
why there is `num <= 0` since the condition is `num > 0`

Reply:
`num` is a decreasing var to record the number of rules that can be added to the group

Reply:
I see. Isn't it better to make it a function? There are similar codes above.

Reply:
It is made now